### PR TITLE
[MRG] Expose original units

### LINKED ISSUE #5607: [Feature Request] expose original units for each channel in BaseRaw
Upon reading a raw data file, MNE-Python converts the units to e.g., Volts (and others, see [here](http://martinos.org/mne/stable/manual/io.html)).

It would be good to expose the channel wise units as they are in the original data file within the `BaseRaw` object.

Concretely, we need this in MNE-BIDS, see: https://github.com/mne-tools/mne-bids/issues/81

### PR DESCRIPTION
**Pull request by sappelhoff**
fixes #5607 

@agramfort see here: This is how I would imagine it ... for now I only did it for BrainVision ... but this scheme could be repeated for the others as well I think.

Opinions?

### CODE REVIEW

**Comment by sappelhoff** [comment]
> One thing we should probably add, too, is ensuring in BaseRaw.__init__ that all units are valid ones from a BIDS perspective.

I agree, however let's do that in another PR. The BIDS-valid units are listed in appendix 5 in the BIDS [specification](http://bids.neuroimaging.io/bids_spec.pdf)

**Comment by larsoner** [comment]
> I agree, however let's do that in another PR.

Okay but let's make sure the original PR and the sanity-checking one are in the same release cycle (so probably 0.18 not 0.17, unless you have time to get both done in the next week)

**Comment by sappelhoff** [comment]
> original PR and the sanity-checking one

I don't follow you there, please help me along:

With this PR, all I want to do is exposing the units as they are in the data files in a private attribute, so that MNE-BIDS can use it for downstream processing.

Apart from that, I agree that in the future, it'd be good if the units handled by MNE-Python would be the same that BIDS allows.

These are two separate issues for me. What do you want to sanity-check?

**Comment by larsoner** [comment]
> These are two separate issues for me. What do you want to sanity-check?

The latter ensures that the former is done correctly. So they are separate / separable if you are willing to accept that the units in PR 1 could be wrong, and that we do not verify it.

**Comment by sappelhoff** [comment]
> The latter ensures that the former is done correctly. So they are separate / separable if you are willing to accept that the units in PR 1 could be wrong, and that we do not verify it.

the units in PR 1 cannot be wrong unless they were wrong in the data to begin with. PR 1 takes the units from the data, and makes them available in `BaseRaw._orig_units`. Before this PR (PR 1), the original units were lost after MNE-Python's IO.

Checking whether the original units are BIDS compatible or not is something that can be done in MNE-BIDS for now, which will do the processing.

**Comment by larsoner** [comment]
> the units in PR 1 cannot be wrong unless they were wrong in the data to begin with.

I'm thinking someone could easily write (and we could easily miss during review) that someone has written `uV` or `μv` instead of `μV`, or failed to put in a unit at all (empty string), etc.

**Comment by sappelhoff** [comment]
Mmh the data that I am primarily targeting is written by e.g., BrainVision recorder ... the professional software sold by Brain Products, who came up with BrainVision.

I think that the issue that you mention applies to data that has been written through custom software or file format conversion software such as https://github.com/palday/philistine/blob/master/philistine/mne/io.py ...

however, again my argument would be that these potential issues should be checked by whoever wants to use `_orig_units`. I think it's fine for MNE-Python to say "look, we read a data file and tell you exactly what's in there ... no more no less."

**Comment by larsoner** [comment]
... but ultimately you (and other BIDS folks) are probably the ones who will have to deal with the `_orig_units` stuff, so if you don't agree with the approach I'm advocating I can live without the sanity checks

**Comment by sappelhoff** [comment]
I think that an `_orig_units` attribute will be helpful for all file formats that are accepted by BIDS. I.e.,

#### MEG
- fif
- ctf
- 4d
- kit
- kdf
- itab
- ... (see appendix 16 in http://bids.neuroimaging.io/bids_spec.pdf)

#### EEG
- BrainVision
- EDF / BDF
- CNT
- EEGLAB

##### iEEG
- MEF3
- NWB

so I think it is reasonable, as you mentioned before, to make `_orig_units` part of `BaseRaw.__init__`

Regarding the other point, I think we are misunderstanding each other ... I'll look at your comments again at a later point. Or perhaps somebody else can clarify such as @jasmainak

File: mne/io/brainvision/brainvision.py [hunk 1]
```diff
@@ -142,7 +143,7 @@ def __init__(self, vhdr_fname, montage=None,
         self._create_event_ch(np.empty((0, 3)), n_samples)
 
         super(RawBrainVision, self).__init__(
-            info, last_samps=[n_samples - 1], filenames=[data_filename],
+            info, last_samps=[n_samples - 1], filenames=[data_fname],
             orig_format=fmt, preload=preload, verbose=verbose,
             raw_extras=[offsets])
```
[the remark this answers is not in the document]
Comment (sappelhoff) [reply mne/io/brainvision/brainvision.py]: for that, would I have to edit the `BaseRaw` class in `base.py`? E.g., by always initializing `_orig_units = dict()` ... and then calling `self._orig_units.update(orig_units)` in the concrete IO handlers (such as RawBrainVision)?

[the remark this answers is not in the document]
Comment (larsoner) [reply mne/io/brainvision/brainvision.py]: Each reader would be responsible for constructing this `dict` before calling `BaseRaw.__init__(..., orig_units=orig_units)`, and all `BaseRaw.__init__` would do is any necessary sanity checking (at the very least, that all `info['ch_names']` entries are dict keys or so) and set `self._orig_units = orig_units`

**Comment by agramfort** [comment]
I would focus on files that are supported by BIDS EEG so edf, eeglab and brainvision.

@sappelhoff see why CIs complain

**Comment by sappelhoff** [comment]
> I would focus on files that are supported by BIDS EEG so edf, eeglab and brainvision.

yes, me too. However, with this PR I'll provide the infrastructure that can then be easily used to expose `_orig_units` for other file formats as well.

> what will raw._orig_units be if the raw object was constructed using RawArray

Currently, I initialize it to `None`, which is fine because it's a private attribute mainly for our internal use and not for end-users ... right?

> @sappelhoff see why CIs complain

Mmmh `mne/tests/test_docstring_parameters.py` failed ... so is that because I didn't write a documentation for `_orig_units`? it's a private attribute though, so where would I describe it?

------------------

finally, regarding my discussion with @larsoner: I still don't understand the points fully - but my suggestion is that I push this PR forward and then get suggestions for sanity checks

**Comment by agramfort** [comment]
run the test locally to replicate the issue.

parameters should be documented in the right order

**Comment by sappelhoff** [comment]
the code is done now for BrainVision and EDF ... EEGLAB doesn't seem to provide units (it's a secondary file format already) ... and for NEUROSCAN CNT, I cannot figure out what to do with sensitivity and calib (see my summary in the issue text #5607).

I would be happy with having BrainVision and EDF covered, which are the recommended file formats for both EEG and iEEG in BIDS.

File: mne/io/base.py [hunk 2]
```diff
@@ -363,6 +368,24 @@ def __init__(self, info, preload=False,
...
+            ch_l = list(info['ch_names'])
+
+            # STI 014 channel is native only to fif ... for all other formats
+            # this was artificially added by the IO procedure
+            if 'STI 014' in ch_l and not self.filenames[0].endswith('.fif'):
+                ch_l.remove('STI 014')
+
+            # check that channels correspond. note that long ch_names get
+            # truncated, so check first 15 characters
+            orig_units_chs = [ch[:15] for ch in orig_units.keys()]
```
Comment (agramfort) [inline mne/io/base.py]: no need for .keys()

Comment (agramfort) [reply mne/io/base.py]: it seems you want to do an intersection between set of ch_l and orig_units. why not using a `set` and use its builtin intersection method?

https://docs.python.org/3/tutorial/datastructures.html#sets
https://docs.python.org/3/library/stdtypes.html#set

File: mne/io/base.py [hunk 3]
```diff
@@ -363,6 +368,24 @@ def __init__(self, info, preload=False,
...
+            ch_correspond = [ch in orig_units_chs for ch in ch_l]
+            if not all(ch_correspond):
+                ch_without_orig_unit = ch_l[ch_correspond.index(False)]
+                raise ValueError('Channel {0} has no associated original '
+                                 'unit.'.format(ch_without_orig_unit))
```
Comment (agramfort) [inline mne/io/base.py]: not covered by test

Comment (sappelhoff) [reply mne/io/base.py]: where should I place the test? There is no `mne/io/tests/test_base.py`

File: mne/io/base.py (hunk 2 above)
Comment (sappelhoff) [reply mne/io/base.py]: the `ch[:15]` part is important to account for truncated channel names, so I need the first list comprehension for this operation ... and then it's just one other line with a list comprehension --> is there a significant benefit to using a set intersection here? Or is it a question of style?

**Comment by agramfort** [comment]
>
> the ch[:15] part is important to account for truncated channel names, so I need the first list comprehension for this operation ... and then it's just one other line with a list comprehension --> is there a significant benefit to using a set intersection here? Or is it a question of style?
>
>
I find it more readable. It took me multiple sections to understand what
you were trying to do.
Using set would be more readable I think.

**Comment by sappelhoff** [comment]
> I find it more readable. It took me multiple sections to understand what you were trying to do.

I tried to use the intersection but I found the list comprehension to be easier actually, would you mind reading over the section again? I restructured a bit and it should be more easily understandable now ... if not, I'll give the set-intersection approach another try.

File: mne/io/brainvision/tests/test_brainvision.py [hunk 4]
```diff
@@ -62,6 +63,28 @@
...
+    assert orig_units['HR'] == 'uS'
+    assert orig_units['Vb'] == 'S'
+    assert orig_units['ReRef'] == 'C'
+
+    # Also test handling in base.py
+    # Should complain that channel Cz does not have a corresponding original
+    # unit.
+    with pytest.raises(ValueError, match='has no associated original unit.'):
+        info = create_info(ch_names=['Cz'], sfreq=100, ch_types='eeg')
+        raw = BaseRaw(info, last_samps=[1], orig_units=['not_Cz'])
```
Comment (sappelhoff) [inline mne/io/brainvision/tests/test_brainvision.py]: @agramfort see the test for covering the part in `base.py`

**Comment by sappelhoff** [comment]
ready for a review @agramfort @larsoner @jasmainak

File: mne/io/base.py [hunk 5]
```diff
@@ -286,6 +287,9 @@ class BaseRaw(ProjMixin, ContainsMixin, UpdateChannelsMixin,
     buffer_size_sec : float
         The buffer size in seconds that should be written by default using
         :meth:`mne.io.Raw.save`.
+    orig_units : dict | None
+        Dictionary mapping channel names to their units as specified in
+        the header file.
```
Comment (jasmainak) [inline mne/io/base.py]: can you add an example?

File: mne/io/base.py [hunk 6]
```diff
@@ -363,6 +368,26 @@ def __init__(self, info, preload=False,
...
         self.orig_format = orig_format
+        # Sanity check and set original units, if provided by the reader:
+        if orig_units:
+            # original units need to be truncated to 15 chars, which is what
+            # the MNE IO procedure also does with the other channels
+            orig_units_trunc = [ch[:15] for ch in orig_units]
+
+            # STI 014 channel is native only to fif ... for all other formats
+            # this was artificially added by the IO procedure, so remove it
+            ch_l = list(info['ch_names'])
```
Comment (jasmainak) [inline mne/io/base.py]: `ch_names` not `ch_l`?

File: mne/io/base.py [hunk 7]
```diff
@@ -363,6 +368,26 @@ def __init__(self, info, preload=False,
         self._comp = None
         self._filenames = list(filenames)
         self.orig_format = orig_format
+        # Sanity check and set original units, if provided by the reader:
+        if orig_units:
```
Comment (jasmainak) [inline mne/io/base.py]: `orig_units` should be same length as `info['ch_names']` ? and the same `ch_names`? do you validate this somewhere?

File: mne/io/brainvision/tests/test_brainvision.py [hunk 8]
```diff
@@ -62,6 +63,28 @@
...
+    assert orig_units['FP1'] == u'µV'
+    assert orig_units['CP5'] == 'BS'
+    assert orig_units['CP6'] == u'µS'
+    assert orig_units['HL'] == 'ARU'
+    assert orig_units['HR'] == 'uS'
+    assert orig_units['Vb'] == 'S'
+    assert orig_units['ReRef'] == 'C'
+
+    # Also test handling in base.py
+    # Should complain that channel Cz does not have a corresponding original
```
Comment (jasmainak) [inline mne/io/brainvision/tests/test_brainvision.py]: channel not_Cz

File: mne/io/base.py (hunk 7 above)
Comment (jasmainak) [inline mne/io/base.py]: `if orig_units is not None` ? Otherwise, this check will go through if I give `orig_units=True`

**Comment by agramfort** [comment]
@sappelhoff good to go from your end?

File: mne/io/brainvision/tests/test_brainvision.py (hunk 8 above)
Comment (sappelhoff) [reply mne/io/brainvision/tests/test_brainvision.py]: no this is actually true ... I initialize a raw object with a channel Cz ... but the corresponding channel in `orig_units` is called not_Cz, so we get a complaint, that for channel Cz, no corresponding original unit was found.

File: mne/io/base.py (hunk 7 above)
Comment (sappelhoff) [reply mne/io/base.py]: thanks! I will raise an error if orig_units is not None and not a dict.

File: mne/io/base.py [hunk 9]
```diff
@@ -363,6 +368,31 @@ def __init__(self, info, preload=False,
...
+                ch_names.remove('STI 014')
+
+            # Each channel in the data must have a corresponding channel in
+            # the original units.
+            ch_correspond = [ch in orig_units_trunc for ch in ch_names]
+            if not all(ch_correspond):
+                ch_without_orig_unit = ch_names[ch_correspond.index(False)]
+                raise ValueError('Channel {0} has no associated original '
+                                 'unit.'.format(ch_without_orig_unit))
+            self._orig_units = orig_units
```
Comment (larsoner) [inline mne/io/base.py]: This should probably not be indented. That way `raw._orig_units` is always there, but could be `None`. This is easier to work with using `if raw._orig_units is not None` than `if hasattr(raw, '_orig_units')` or a `try/except`

File: mne/io/base.py (hunk 7 above)
Comment (sappelhoff) [reply mne/io/base.py]: this is a bit tricky ... I make sure that all channels and units read from the raw file are in `orig_units`. But the MNE IO procedure changes a few things on the way, e.g.:

1. adding a stim channel
2. truncating long channel names so that they have max. 15 chars
3. adding or removing some other channels depending on the IO arguments the user provides (excluding, etc.)

So bottom line: currently I only check that each channel in `info['ch_names']` does have a corresponding key in `orig_units`, which I think is what we want, and thus it's fine.

**Comment by sappelhoff** [comment]
> @sappelhoff good to go from your end?

from my end yes, as soon as CIs are green. There is one comment by @jasmainak that I did not address, see: https://github.com/mne-tools/mne-python/pull/5633#discussion_r227365746

so it depends on how my response is evaluated by you guys, if valid - then let's merge.

**Comment by sappelhoff** [comment]
> This ensures that all raw readers, if they supply _orig_units, give valid ones. Eventually hopefully we could get rid of the conditional if raw._orig_units is not None since ideally all readers would give the units (right?), and ensures MNE is giving something sensible. WDYT?

I fully agree ...

> ... assuming you can construct the valid_units set in some loop somewhere.

... but I am at a loss how you want me to construct `valid_units` or where to get them from --> this relates to our previous discussion yesterday.

File: mne/io/brainvision/tests/test_brainvision.py [hunk 10]
```diff
@@ -62,6 +63,33 @@
...
+    # unit.
+    with pytest.raises(ValueError, match='has no associated original unit.'):
```
Comment (sappelhoff) [inline mne/io/brainvision/tests/test_brainvision.py]: also, would you know a better place for this test and the one on line 88 @larsoner?

both of them are general and not BrainVision specific ... they need to be run only once (so probably they should not be in `test_raw.py::_test_raw_reader`), but they could just as well be in EDF - or whatever file format that we expose the orig_units for in the future

**Comment by larsoner** [comment]
> ... but I am at a loss how you want me to construct valid_units or where to get them from --> this relates to our previous discussion yesterday.

We should use / construct the BIDS-valid list

File: mne/io/brainvision/tests/test_brainvision.py (hunk 10 above)
Comment (larsoner) [reply mne/io/brainvision/tests/test_brainvision.py]: You can create a new `io/tests/test_base.py` if you want

**Comment by sappelhoff** [comment]
> We should use / construct the BIDS-valid list

okay - but the manufacturer (say Brain Products) don't necessarily adhere to the units that are BIDS-valid, and thus they will have units in their raw data, that will fail our test --> how should we deal with that?

**Comment by larsoner** [comment]
> okay - but the manufacturer (say Brain Products) don't necessarily adhere to the units that are BIDS-valid, and thus they will have units in their raw data, that will fail our test --> how should we deal with that?
>
> edit: Or are you talking about minor things such as specifying microVolts as "uV" or "µV", which we could have a conversion dictionary for?

Yes we should do the work at the MNE end to make sure entries in `raw._orig_units` are consistent. I see the job of `mne.io` to provide convenient abstractions and consistent interfaces. This means each reader is responsible for having `raw._orig_units` have entries that are BIDS valid. In a worst case this should be `Unknown` (or whatever it's called in BIDS, there is such a thing right?) if e.g. BrainVision or some other manufacturer starts adding things that are not supported.

File: mne/io/meas_info.py [hunk 11]
```diff
@@ -53,6 +53,9 @@
     hbr=(FIFF.FIFFV_FNIRS_CH, FIFF.FIFFV_COIL_FNIRS_HBR, FIFF.FIFF_UNIT_MOL)
 )
 
+# Valid units to be exposed in raw._orig_units
+valid_units = ['uV', u'µV']
```
Comment (sappelhoff) [inline mne/io/meas_info.py]: work in progress, depending on how discussion with @larsoner goes

**Comment by larsoner** [comment]
Eventually such an effort can be integrated with the FIF units:

https://github.com/mne-tools/mne-python/blob/master/mne/io/constants.py#L706

Maybe the best way forward come to think of it would be to always use `info['chs'][ii]['unit']`, converted to a `str`.

**Comment by larsoner** [comment]
... oh never mind that last comment, **orig** units will be different from `info['chs'][ii]['unit']` which can be scaled

**Comment by sappelhoff** [comment]
Ahhh now I see your point, thanks for bearing with me for so long.

> In a worst case this should be Unknown (or whatever it's called in BIDS, there is such a thing right?) if e.g. BrainVision or some other manufacturer starts adding things that are not supported.

in BIDS, that would be `n/a`.

So I will try to come up with a private function that checks units and tries to convert them to BIDS valid units ... if this does not work, I will assign `n/a`.

File: mne/io/base.py (hunk 7 above)
Comment (jasmainak) [reply mne/io/base.py]: okay that's fine then

File: mne/io/utils.py [hunk 12]
```diff
@@ -13,6 +14,50 @@
...
+    for ch_name, unit in orig_units_remapped.items():
+
+        # Be lenient: we ignore case for now.
+        if unit.lower() in valid_units_lowered:
+            continue
+
+        # Common "invalid units" can be remapped to their valid equivalent
+        if unit.lower() == 'uv':
+            orig_units_remapped[ch_name] = u'µV'
+            continue
```
[the remark this answers is not in the document]
Comment (larsoner) [reply mne/io/utils.py]: I would only add ones that we know we need. Assume YAGNI for others (including "seconds" to "second", unless you've already seen it in a file)

**Comment by sappelhoff** [comment]
The PR is good to go from my side @agramfort @larsoner @jasmainak 

one strange CI error remains: `E   AttributeError: module 'matplotlib' has no attribute 'artist'`

it seems unrelated ...

File: mne/io/brainvision/brainvision.py [hunk 13]
```diff
@@ -141,10 +142,13 @@ def __init__(self, vhdr_fname, montage=None,
         # Create a dummy event channel first
         self._create_event_ch(np.empty((0, 3)), n_samples)
 
+        # Final check of orig_units, editing a unit if it is not a valid unit
+        orig_units = _check_orig_units(orig_units)
```
Comment (larsoner) [inline mne/io/brainvision/brainvision.py]: Shouldn't this be in `BaseRaw.__init__()`?

File: mne/io/edf/edf.py [hunk 14]
```diff
@@ -192,11 +196,15 @@ def __init__(self, input_fname, montage, eog=None, misc=None,
             warn("'annot' and 'annotmap' parameters are deprecated and will be"
                  " removed in 0.18", DeprecationWarning)
 
+        # Final check of orig_units, editing a unit if it is not a valid unit
+        orig_units = _check_orig_units(orig_units)
```
Comment (larsoner) [inline mne/io/edf/edf.py]: Same, shouldn't be here but in `BaseRaw.__init__`

File: mne/io/meas_info.py [hunk 15]
```diff
@@ -53,6 +54,44 @@
...
+                 for unit in valid_unit_names])
+valid_units += ([''.join([prefix, unit]) for prefix in valid_prefix_symbols
+                 for unit in valid_unit_symbols])
+
+# units are also valid without a prefix
+valid_units += valid_unit_names
+valid_units += valid_unit_symbols
+
+# we also accept "n/a" as a unit, which is the default missing value in BIDS
+valid_units += ["n/a"]
```
Comment (larsoner) [inline mne/io/meas_info.py]: I am a bit wary of putting things like this in the file itself because it adds to import time. How about a `get_valid_units()` function?

Also it should probably return a `tuple` not `list` so it's immutable

File: mne/io/base.py [hunk 16]
```diff
@@ -363,6 +369,35 @@ def __init__(self, info, preload=False,
...
+            # original units need to be truncated to 15 chars, which is what
+            # the MNE IO procedure also does with the other channels
+            orig_units_trunc = [ch[:15] for ch in orig_units]
+
+            # STI 014 channel is native only to fif ... for all other formats
+            # this was artificially added by the IO procedure, so remove it
+            ch_names = list(info['ch_names'])
+            if ('STI 014' in ch_names) and not \
+               (self.filenames[0].endswith('.fif')):
+                ch_names.remove('STI 014')
```
Comment (massich) [inline mne/io/base.py]: I don't understand this. Why are we removing the stim channel here? wouldn't this have collateral effects? I would let it be until we remove it for good.

File: mne/io/brainvision/brainvision.py [hunk 17]
```diff
@@ -956,7 +968,8 @@ def _get_vhdr_info(vhdr_fname, eog, misc, scale, montage):
 
     info._update_redundant()
     info._check_consistency()
-    return info, data_filename, fmt, order, mrk_fname, montage, n_samples
+    return (info, data_fname, fmt, order, n_samples, mrk_fname, montage,
+            orig_units)
```
Comment (massich) [inline mne/io/brainvision/brainvision.py]: why did the order of the parameters change?

File: mne/io/base.py (hunk 16 above)
Comment (sappelhoff) [reply mne/io/base.py]: it is only being removed for the sake of the check, it stays in the data after the check.

for the check however, it needs to be removed because we do not have an "original unit" for the stim channel if it was artificially added during the IO procedure (which is always the case except for fif)

File: mne/io/brainvision/brainvision.py [hunk 18]
```diff
@@ -648,7 +658,7 @@ def _get_vhdr_info(vhdr_fname, eog, misc, scale, montage):
 
     # locate EEG binary file and marker file for the stim channel
     path = op.dirname(vhdr_fname)
-    data_filename = op.join(path, cfg.get(cinfostr, 'DataFile'))
+    data_fname = op.join(path, cfg.get(cinfostr, 'DataFile'))
```
Comment (massich) [inline mne/io/brainvision/brainvision.py]: why are changing the variable name? I mean, I'm ok with both. Do we have a naming convention or could we start one?

File: mne/io/brainvision/brainvision.py (hunk 17 above)
Comment (sappelhoff) [reply mne/io/brainvision/brainvision.py]: because I corrected a copy paste error and changed some names in the process, see https://github.com/mne-tools/mne-python/pull/5633/files#r226644804

File: mne/io/brainvision/brainvision.py (hunk 18 above)
Comment (sappelhoff) [reply mne/io/brainvision/brainvision.py]: this was for two reasons:
1. it's shorter
2. it's consistent with `marker_fname`, which we were using before

File: mne/io/brainvision/tests/test_brainvision.py [hunk 19]
```diff
@@ -62,6 +63,21 @@
 event_id = {'Sync On': 5}
 
 
+def test_orig_units():
+    """Test exposure of original channel units."""
+    with pytest.warns(RuntimeWarning, match='will be dropped'):
```
Comment (massich) [inline mne/io/brainvision/tests/test_brainvision.py]: In this case since the warning is unrelated and we are not testing any other warning I would use pytest's fixture `reacwarn`.

```py
def test_orig_units(recwarn):
    orig_units = read_raw_brainvision(vhdr_path)._orig_units
	assert len(orig_units) == 32
￼    assert orig_units['FP1'] == u'µV'
	...
```
￼

File: mne/io/brainvision/tests/test_brainvision.py [hunk 20]
```diff
@@ -62,6 +63,21 @@
...
+        raw = read_raw_brainvision(vhdr_path)
```
Comment (massich) [inline mne/io/brainvision/tests/test_brainvision.py]: how much does the test takes to run? maybe we should start thinking on using a fixture to load the data and do it only once for the entire testsuit, and save some CI cycles.

File: mne/io/edf/tests/test_edf.py [hunk 21]
```diff
@@ -123,6 +125,12 @@ def test_edf_data():
                            stim_channel=None, exclude=['Ergo-Left', 'H10'],
                            verbose='error')
     raw_py = read_raw_edf(edf_path, stim_channel='auto', preload=True)
+
+    # Test original units
```
Comment (massich) [inline mne/io/edf/tests/test_edf.py]: as in brainvision if we load the file in anyother place without using `_test_raw_reader` i would factor it out, use a fixutre and create a **unit** test for the units. Otherwise, add it to the common tests in `_test_raw_reader` but I think that the later wont be easy.

File: mne/io/brainvision/tests/test_brainvision.py (hunk 20 above)
Comment (sappelhoff) [reply mne/io/brainvision/tests/test_brainvision.py]: it's pretty fast:
```
$ pytest mne/io/brainvision/tests/test_brainvision.py::test_orig_units --durations=0
Test session starts (platform: linux, Python 3.6.6, pytest 3.7.3, pytest-sugar 0.9.1)

 mne/io/brainvision/tests/test_brainvision.py ✓                                                                 100% ██████████
=================================================== slowest test durations ====================================================
0.01s call     mne/io/brainvision/tests/test_brainvision.py::test_orig_units
0.00s setup    mne/io/brainvision/tests/test_brainvision.py::test_orig_units
0.00s teardown mne/io/brainvision/tests/test_brainvision.py::test_orig_units

Results (0.20s):
       1 passed


```

File: mne/io/brainvision/tests/test_brainvision.py (hunk 19 above)
Comment (sappelhoff) [reply mne/io/brainvision/tests/test_brainvision.py]: I don't know about reacwarn, could you elaborate or share a link? I can't find something on Google

File: mne/io/edf/tests/test_edf.py (hunk 21 above)
Comment (sappelhoff) [reply mne/io/edf/tests/test_edf.py]: what's the argument against having the test here? It does not require additional loading, simply accessing an existing attribute and making two logical comparisons

File: mne/io/meas_info.py [hunk 22]
```diff
@@ -54,6 +55,60 @@
 )
 
 
+def _get_valid_units():
```
Comment (massich) [inline mne/io/meas_info.py]: I'm not sure if we should have this here buried in our code or somewhere else with 'external'. If we had a dependency in BIDS I would rather like to do  `from bids import VALID_UNITS` which they export as a constant. Since thats not the case I would like it better to do `from mne.externals.bids import VALID_UNITS`

File: mne/io/tests/test_utils.py [hunk 23]
```diff
@@ -0,0 +1,19 @@
...
+# Author: Stefan Appelhoff <stefan.appelhoff@mailbox.org>
+#
+# License: BSD (3-clause)
+
+from ..utils import _check_orig_units
+
+
+def test_check_orig_units():
+    """Test the checking of original units."""
+    orig_units = dict(FC1='nV', Hfp3erz='n/a', Pz='uV', greekMu=u'μV',
```
Comment (massich) [inline mne/io/tests/test_utils.py]: should we test for wrong capitalization? do we consider them valid or invalid?

Comment (sappelhoff) [reply mne/io/tests/test_utils.py]: for now I suggest that we are lenient and accept any case. This can be changed once BIDS also employs stronger tests ... so far, the https://github.com/bids-standard/bids-validator/ does not check units for validity, so we are already a step ahead.

File: mne/io/meas_info.py (hunk 22 above)
Comment (sappelhoff) [reply mne/io/meas_info.py]: These units are actually the international units ("SI" units): https://en.wikipedia.org/wiki/International_System_of_Units

**Comment by agramfort** [comment]
@sappelhoff let me know when you want me to a final review.

@massich feel free to propose commits to @sappelhoff so he can integrate them. Otherwise we'll do it in a subsequent PR.

**Comment by sappelhoff** [comment]
> @sappelhoff let me know when you want me to a final review.

ready now :-)

File: mne/io/brainvision/tests/test_brainvision.py (hunk 19 above)
Comment (massich) [reply mne/io/brainvision/tests/test_brainvision.py]: Sorry, it was a typo. But the snipped is correct. Just add `recwarn` as a test parameter, and the whole thing works.

```diff
- def test_orig_units():
+ def test_orig_units(recwarn):
```
```diff
-    with pytest.warns(RuntimeWarning, match='will be dropped'):
-        raw = read_raw_brainvision(vhdr_path)
-    orig_units = raw._orig_units
+    orig_units = read_raw_brainvision(vhdr_path)._orig_units
```

File: mne/io/brainvision/brainvision.py (hunk 17 above)
Comment (massich) [reply mne/io/brainvision/brainvision.py]: +1

File: mne/io/base.py (hunk 16 above)
Comment (massich) [reply mne/io/base.py]: shouldn't it then treat in the check? rather than modifying the raw object?

File: mne/io/edf/tests/test_edf.py (hunk 21 above)
Comment (massich) [reply mne/io/edf/tests/test_edf.py]: That brainvision has a `test_original_units()` and edf does not. To me, all readers should have similar tests and structured in the same manner. Then if the argument of just sneaking couple asserts here with a comment `# Test original units` instead of defining a function `def test_original_utings()` is to avoid overhead, we should start fixtures. Even if they are pretty fast.

File: mne/io/base.py (hunk 16 above)
Comment (sappelhoff) [reply mne/io/base.py]: I don't know what you mean, could you provide an example? Where am I modifying a raw object here?

Comment (sappelhoff) [reply mne/io/base.py]: I am working on a copy of the ch_names, see line 384

Comment (massich) [reply mne/io/base.py]: yes, you are completely right. you are working with a copy. my bad.

**Comment by massich** [comment]
@sappelhoff I've restarted the CIs

File: mne/io/tests/test_base.py [hunk 24]
```diff
@@ -0,0 +1,28 @@
...
+    # Should complain that channel Cz does not have a corresponding original
+    # unit.
+    with pytest.raises(ValueError, match='has no associated original unit.'):
+        info = create_info(ch_names=['Cz'], sfreq=100, ch_types='eeg')
+        BaseRaw(info, last_samps=[1], orig_units={'not_Cz': 'nV'})
+
+    # Test that a non-dict orig_units argument raises a ValueError
+    with pytest.raises(ValueError, match='orig_units must be of type dict'):
+        info = create_info(ch_names=['Cz'], sfreq=100, ch_types='eeg')
+        BaseRaw(info, last_samps=[1], orig_units=True)
```
Comment (agramfort) [inline mne/io/tests/test_base.py]: on a second though I think this should be put in this file:

https://github.com/mne-tools/mne-python/blob/f81bf9b96c7e530148e8781357d9e8a71fc9a418/mne/io/tests/test_raw.py

and I would update the common raw test:

https://github.com/mne-tools/mne-python/blob/f81bf9b96c7e530148e8781357d9e8a71fc9a418/mne/io/tests/test_raw.py#L24:5

so you have a test that orig_units is present for all raw readers.

Comment (sappelhoff) [reply mne/io/tests/test_base.py]: so you suggest that I migrate this code from `test_base.py` into the function `_test_raw_reader` within `test_raw.py`?

Or that I create a new test function within `test_raw.py`?

**Comment by agramfort** [comment]
I would move the function to test_raw.py and update the _test_raw_reader
function to make sure the orig_units are proper for all readers ie present
and valid or None.

**Comment by massich** [comment]
do you really think that this should be in the common test, already? I think we should first add it and then integrate it to the common test if we support it for all readers.

**Comment by agramfort** [comment]
at least check that attribute is always present even if it's None

**Comment by massich** [comment]
if that's the case, then it should truly be in the common test. Then I would like to think of how to split the common testing. But this can be done further on.

**Comment by larsoner** [comment]
Agreed it should be in `_test_raw_reader`, None is okay for now, eventually we will eliminate it

File: mne/io/tests/test_base.py (hunk 24 above)
Comment (sappelhoff) [reply mne/io/tests/test_base.py]: @agramfort @larsoner @massich I don't really understand what you are planning. The tests here in `test_base.py` 
should be moved somewhere else?

1. Why? As far as I see it the tests here make sure that functionality in base.py is working properly, so it seemed like a logical place. I would keep it here and simply add a line `assert hasattr(raw, '_orig_units')` in `_test_raw_reader`

2. Where - you say to `test_raw.py`?

3. How - in a new `test_orig_units` function within `test_raw.py` ... or as part of `_test_raw_reader` within `test_raw.py`

File: mne/io/tests/test_raw.py [hunk 25]
```diff
@@ -110,6 +117,15 @@ def _test_raw_reader(reader, test_preloading=True, **kwargs):
...
     assert isinstance(raw.annotations, Annotations)
 
+    # Make a "soft" test on units: They have to be valid SI units as in
+    # mne.io.meas_info.valid_units, but we accept any lower/upper case for now.
+    valid_units = _get_valid_units()
+    valid_units_lower = [unit.lower() for unit in valid_units]
+    if raw._orig_units is not None:
+        assert isinstance(raw._orig_units, dict)
+        for ch_name, unit in raw._orig_units.items():
+            assert unit.lower() in valid_units_lower, ch_name
```
Comment (larsoner) [inline mne/io/tests/test_raw.py]: Ahh you already have a test in `_test_raw_reader`, I think the PR is okay as-is

File: mne/io/tests/test_base.py (hunk 24 above)
Comment (larsoner) [reply mne/io/tests/test_base.py]: Looking closer I disagree @agramfort -- this test looks at corner cases of `BaseRaw.__init__`, so this is a reasonable place (and should only be run once!). @sappelhoff already has tests that `_orig_units` is present and is `None` or `dict` below in `_test_raw_reader` so all readers are tested.

Eventually we should be able to get rid of the `if raw._orig_units is not None:` conditional, but not until we extend it to all readers

Comment (larsoner) [reply mne/io/tests/test_base.py]: ... ahh @sappelhoff I think maybe all you need to do (since you already have the `_test_raw_reader` tests, @agramfort probably just missed them) is move/copy-paste this test function to `mne/io/tests/test_raw.py` rather than creating this new file. Sorry I led you astray on that!

**Comment by larsoner** [comment]
> Agreed it should be in _test_raw_reader, None is okay for now, eventually we will eliminate it

I take this back -- to me it looks like what should be in `test_raw.py` is already there, and all that is now required is the content from `mne/io/tests/test_base.py` to `mne/io/tests/test_raw.py`

**Comment by sappelhoff** [comment]
@larsoner see https://github.com/mne-tools/mne-python/pull/5633/commits/159677f28d7217cc4ebc57958ee21b43903749aa
